src/include/buffer.h

```diff
@@ -594,13 +594,13 @@ struct error_code;
 	}
       }
       void clear_and_dispose() {
-	for (auto it = begin(); it != end(); /* nop */) {
-	  auto& node = *it;
-	  it = it->next;
-	  ptr_node::disposer()(&node);
-	}
-	_root.next = &_root;
-	_tail = &_root;
+        auto it = begin();
+        while (!empty() || it != end()) {
```

Comment:
why do we need to check for `empty()` in each iteration? comparing `it` with `end()` would suffice. and neither do i think we need to create a temporary iterator in each iteration. so probably we can have:
```c++
for (auto it = begin(), e = end(); it != e; ++it) {
  auto& node = *it;
  ptr_node::disposer()(&node);
}
_root.next = &_root;
_tail = &_root;
```

sorry for the bike-shedding, after a second thought, i think we can probably improve the code a little bit further.

Reply:
> why do we need to check for `empty()` in each iteration?

According to the tracker, we have gotten to a situation where both `it->cur` and `this->_root.next` is null. I was suspecting that we are are not correctly handing an empty lists (Can't verify yet).

>  and neither do i think we need to create a temporary iterator in each iteration.
We need to, otherwise we hit a segfault.

Thank you for the quick response.



If we won't create a temporary iterator we will hit a segfault.

Reply:
@Matan-B sorry for the confusion, i meant to refer `end()`, which creates a temporary `iterator` instance.